File: server/src/shared/user-context.ts
import { AsyncLocalStorage } from 'node:async_hooks';
import type { FastifyRequest } from 'fastify';
import { loadServerConfig } from './config';
import { verifySessionToken, isSignedToken } from './session';

export const USER_ID_HEADER = 'x-user-id';
export const DEV_SESSION_COOKIE = 'table_dev_session_user_id';

export type ServerUserContext = {
  userId: string;
  source: 'default' | 'header' | 'signed_session' | 'missing';
};

const userContextStorage = new AsyncLocalStorage<ServerUserContext>();

export function getDefaultUserId() {
  return loadServerConfig().DEFAULT_USER_ID;
}

function readCookieValue(request: FastifyRequest, cookieName: string): string | null {
  const rawCookie = request.headers.cookie;
  if (!rawCookie) {
    return null;
  }

  const cookieParts = rawCookie.split(';');
  for (const part of cookieParts) {
    const [name, ...valueParts] = part.trim().split('=');
    if (name !== cookieName) {
      continue;
    }

    const value = valueParts.join('=').trim();
    if (!value) {
      return null;
    }

    try {
      return decodeURIComponent(value);
    } catch {
      return value;
    }
  }

  return null;
}

export function resolveRequestUserContext(request: FastifyRequest): ServerUserContext {
  const config = loadServerConfig();

  // 1. 检查签名 Cookie（最高优先级）
  const cookieValue = readCookieValue(request, DEV_SESSION_COOKIE);
  if (cookieValue && isSignedToken(cookieValue)) {
    const verifiedUserId = verifySessionToken(cookieValue);
    if (verifiedUserId) {
      return { userId: verifiedUserId, source: 'signed_session' };
    }
    // 签名无效或过期，继续尝试其他方式
  }

  // 2. 检查 x-user-id 头（受 TRUST_USER_ID_HEADER 控制）
  const headerValue = request.headers[USER_ID_HEADER];
  const hasHeader = typeof headerValue === 'string' && headerValue.trim().length > 0;
  if (hasHeader && config.TRUST_USER_ID_HEADER) {
    return { userId: headerValue.trim(), source: 'header' };
  }

  // 3. 回退到默认用户
  return { userId: getDefaultUserId(), source: 'missing' };
}

export function runWithUserContext<T>(context: ServerUserContext, callback: () => T) {
  return userContextStorage.run(context, callback);
}

export function getCurrentUserContext(): ServerUserContext {
  return userContextStorage.getStore() ?? {
    userId: getDefaultUserId(),
    source: 'missing',
  };
}

export function getCurrentUserId() {
  return getCurrentUserContext().userId;
}

File: python-backend/app/core/user_context.py
from contextvars import ContextVar, Token
from dataclasses import dataclass
from typing import Literal
from urllib.parse import unquote
from uuid import UUID

from app.core.config import Settings
from app.core.session import is_signed_token, verify_session_token
from fastapi import Request

UserSource = Literal["default", "header", "signed_session", "missing"]


@dataclass(frozen=True)
class UserContext:
    user_id: str
    source: UserSource


_user_context: ContextVar[UserContext | None] = ContextVar("user_context", default=None)
USER_ID_HEADER = "x-user-id"
DEV_SESSION_COOKIE = "table_dev_session_user_id"


def _read_cookie_value(request: Request, cookie_name: str) -> str | None:
    value = request.cookies.get(cookie_name)
    if not value:
        return None
    try:
        return unquote(value)
    except Exception:
        return value


def resolve_request_user_context(request: Request, settings: Settings) -> UserContext:
    cookie_value = _read_cookie_value(request, DEV_SESSION_COOKIE)
    if cookie_value and is_signed_token(cookie_value):
        verified_user_id = verify_session_token(cookie_value, settings.provider_secret_key)
        if verified_user_id:
            return UserContext(user_id=verified_user_id, source="signed_session")

    header_value = request.headers.get(USER_ID_HEADER)
    if header_value and settings.trust_user_id_header:
        return UserContext(user_id=header_value.strip(), source="header")
    return UserContext(user_id=settings.default_user_id, source="missing")


def set_user_context(context: UserContext) -> Token[UserContext | None]:
    return _user_context.set(context)


def reset_user_context(token: Token[UserContext | None]) -> None:
    _user_context.reset(token)


def get_user_context(settings: Settings) -> UserContext:
    return _user_context.get() or UserContext(user_id=settings.default_user_id, source="missing")


def validate_user_id(raw_user_id: str) -> None:
    UUID(raw_user_id)
